### ISSUE TITLE
ttnn.sum returns inf/non-NaN when input contains NaN (exposed by logit accuracy fix #46471)

### ISSUE THREAD

**Issue by hkwonTT** [tenstorrent/tt-metal#47494]
## Problem Description

After commit [`2d254d5b38e` \"Improve ttnn.logit Accuracy (#46471)\"](https://github.com/tenstorrent/tt-metal/commit/2d254d5b38e42963b73be4e2e47551676a03f246) was merged, a regression was observed in the `tt-mlir` golden test for `stablehlo.batch_norm_grad` (PR [tenstorrent/tt-mlir#8840](https://github.com/tenstorrent/tt-mlir/pull/8840)):

```
FAILED test/python/golden/test_stablehlo_ops.py::test_batch_norm_grad_op[module_batch_norm_grad-ttnn]
actual_pcc=0.0235 < expected_pcc=0.99
```

Investigation revealed a **pre-existing bug in `ttnn.sum`**: when the input tensor contains `NaN` values, `ttnn.sum` returns `inf` or large finite numbers instead of `NaN`, violating IEEE 754.

---

### 1. What changed in #46471

`calculate_sfpu_binary_div` in `ckernel_sfpu_binary.h` now applies a Markstein residual correction for FP32 division accuracy:

```cpp
// Before: single reciprocal-multiply
sfpi::vFloat result = in0 * sfpu_reciprocal_iter<2>(in1);

// After: Markstein refinement
sfpi::vFloat r = sfpu_reciprocal_iter<2>(in1);
sfpi::vFloat result = in0 * r;
if constexpr (is_fp32_dest_acc_en) {
    v_if(sfpi::exexp(result, ...) != 255) {
        sfpi::vFloat e = in0 - result * in1;  // residual
        result = result + e * r;               // correction
    }
    v_endif;
}
```

This changed the behavior of `x / NaN`:

| | Behavior |
|---|---|
| Before | `recip(NaN) ~= 0`, so `x * 0 = 0.0` (incorrect per IEEE 754) |
| After | Markstein residual: `e = x - 0*NaN = NaN`, then `result + NaN*0 = NaN` (correct per IEEE 754) |

The Markstein fix is **mathematically correct**. The division now properly propagates NaN.

### 2. Pre-existing bug in `ttnn.sum` exposed

With the old division behavior (`x / NaN = 0`), NaN was silently suppressed before reaching `ttnn.sum`. After the fix, NaN propagates through division, and reaches `ttnn.sum` -- where `ttnn.sum` **fails to propagate NaN correctly**:

```
ttnn.sum(tensor_containing_NaN) --> inf or large finite values
                                    (expected: NaN per IEEE 754)
```

---

## Minimal Reproduction

```python
import ttnn
import torch

device = ttnn.open_device(device_id=0)

# Input tensor full of NaN
t = torch.full([32, 64], float('nan'))
tt_t = ttnn.from_torch(t, device=device, dtype=ttnn.DataType.Float32, layout=ttnn.TILE_LAYOUT)

result = ttnn.sum(tt_t, dim=0, keepdim=False)
result_cpu = ttnn.to_torch(result)

# Expected: all NaN   (IEEE 754: NaN + x = NaN for all x)
# Actual:   inf or large finite numbers
print(result_cpu)

ttnn.close_device(device)
```

---

## Impact

The `stablehlo.batch_norm_grad` lowering in `tt-mlir` decomposes to:

```
stddev = sqrt(variance + epsilon)            # NaN if variance < -epsilon
normalized = centered_operand / stddev       # Before fix: 0.0; After fix: NaN
grad_scale = sum(grad_output * normalized)   # Before: sum(0)=0; After: sum(NaN)=inf (bug)
```

When the test accidentally supplied negative variance (mathematically invalid input for batch norm), the tt-mlir test was silently passing before #46471 because `x / NaN = 0` masked the invalid input. After #46471, the NaN propagates correctly through division, reaches `ttnn.sum`, and the `ttnn.sum` NaN bug causes `inf` output -- producing a huge mismatch against the CPU golden.

The `tt-mlir` test has been fixed separately to use valid non-negative variance values. The underlying `ttnn.sum` NaN propagation bug remains.

---

## Hardware

Wormhole B0 (tested on n300, originally reported on n150)

---

## Notes

- The Markstein correction in `ckernel_sfpu_binary.h` is **correct** and should be kept.
- This issue tracks the `ttnn.sum` NaN propagation bug that was latent before #46471.
- A similar issue may exist for other reduction ops (`ttnn.mean`, `ttnn.prod`, etc.) -- not yet verified.

**Comment by hkwonTT** [comment]
AI created this issue in the wrong repository.
